### PR TITLE
[CORE-1750] Support user spec in pachctl edit pipeline

### PR DESCRIPTION
**Pull request by robert-uhl**

### CODE REVIEW

File: src/internal/ppsutil/decoder.go [hunk 1]
```diff
@@ -127,3 +131,182 @@ func (r *PipelineManifestReader) NextCreatePipelineRequest() (*ppsclient.CreateP
...
+		return "", errors.New("no pipeline `name` specified")
+	}
+
+	if req.Transform != nil && req.Transform.Image != "" {
+		if !strings.Contains(req.Transform.Image, ":") {
+			fmt.Fprintf(os.Stderr,
+				"WARNING: please specify a tag for the docker image in your transform.image spec.\n"+
+					"For example, change 'python' to 'python:3' or 'bash' to 'bash:5'. This improves\n"+
+					"reproducibility of your pipelines.\n\n")
+		} else if strings.HasSuffix(req.Transform.Image, ":latest") {
```
Comment (acohen4) [inline src/internal/ppsutil/decoder.go:228]: It feels kind of weird to me that there's one particular tag that we attach semantics and a warning to. do we already warn about this? Just my two cents

File: src/internal/ppsutil/decoder.go [hunk 2]
```diff
@@ -127,3 +131,182 @@ func (r *PipelineManifestReader) NextCreatePipelineRequest() (*ppsclient.CreateP
...
+					"'bash:latest' to 'bash:5'. This improves reproducibility of your pipelines.\n\n")
+		}
+	}
+	return string(b), nil
+}
+
+// yamlToJSON converts a YAML node into an any value which represents a JSON
+// value.  Strings are strings; numbers are json.Number; booleans are bools;
+// nulls are nil; arrays are []any; and maps are map[string]any.
+func yamlToJSON(n *yaml.Node) (any, error) {
```
Comment (acohen4) [inline src/internal/ppsutil/decoder.go:241]: all these helpers look like standard lib stuff that you could do with serde or go-yaml. Do we need this custom implementation?

File: src/server/pps/cmds/cmds.go [hunk 3]
```diff
@@ -1797,34 +1817,74 @@ func pipelineHelper(ctx context.Context, pachctlCfg *pachctl.Config, reprocess b
...
 		}
 
-		if request.Pipeline.Project.GetName() == "" {
-			request.Pipeline.Project = &pfs.Project{Name: project}
+		pipeline, ok := spec["pipeline"].(map[string]any)
+		if !ok {
+			return errors.New("must specify pipeline, an object")
+		}
+		project, ok := pipeline["project"]
+		if !ok {
```
Comment (acohen4) [inline src/server/pps/cmds/cmds.go:1839]: all of these checks seem like they should be handled by decoding/encoding

**Review by acohen4 — COMMENTED** [review]
I'm struggling a bit to understand the purpose behind some of the changes to the abstractions. Answering my specific comments should help

File: src/internal/ppsutil/decoder.go [hunk 4]
```diff
@@ -127,3 +131,182 @@ func (r *PipelineManifestReader) NextCreatePipelineRequest() (*ppsclient.CreateP
 		return r.convertRequest(document)
 	}
 }
+
+// A SpecReader reads JSON or YAML specs.  There may be a single spec, a series
+// of YAML documents and/or an array of specs.  This behavior is copied from
+// PipelineManifestReader.
+type SpecReader struct {
```
[the remark this answers is not in the document]
Comment (robert-uhl) [reply src/internal/ppsutil/decoder.go]: Until INT-1006 is done. I’ve deleted all the other instances.

File: src/internal/ppsutil/decoder.go (hunk 1 above)
Comment (robert-uhl) [reply src/internal/ppsutil/decoder.go]: Yes, this was an existing validity check in `PipelineManifestReader`.

File: src/server/pps/cmds/cmds.go (hunk 3 above)
Comment (robert-uhl) [reply src/server/pps/cmds/cmds.go]: The values can be nil, or (in the YAML) of different types.

**Review by acohen4 — APPROVED** [review]
LGTM

**Pull request merged**